geomesa-memory/src/main/java/org/locationtech/geomesa/memory/cqengine/index/GeoIndex.java

```diff
@@ -0,0 +1,194 @@
+    public void init(ObjectStore<O> objectStore, QueryOptions queryOptions) {
+        addAll(ObjectSet.fromObjectStore(objectStore, queryOptions), queryOptions);
+    }
+
+    @Override
+    public boolean addAll(ObjectSet<O> objectSet, QueryOptions queryOptions) {
+        try {
+            boolean modified = false;
+
+            for (O object : objectSet) {
+                Envelope env = ((Geometry)object.getDefaultGeometry()).getEnvelopeInternal();
+                //System.out.println("Adding " + object + " with envelope " + env);
+                index.insert(env, object);
+                modified = true;
+            }
+
+            return modified;
+        }
+        finally {
+            objectSet.close();
+        }
+    }
+
+    @Override
+    public boolean removeAll(ObjectSet<O> objectSet, QueryOptions queryOptions) {
+        try {
+            boolean modified = false;
+
+            for (O object : objectSet) {
+                Envelope env = ((Geometry)object.getDefaultGeometry()).getEnvelopeInternal();
+                index.remove(env, object);
+                modified = true;
+            }
+
+            return modified;
+        }
+        finally {
+            objectSet.close();
+        }
+    }
+
+    @Override
+    public void clear(QueryOptions queryOptions) {
+        this.index = new BucketIndex<SimpleFeature>(360, 180, new Envelope(-180.0, 180.0, -90.0, 90.0));
+    }
+
+    @Override
+    public ResultSet<O> retrieve(final Query<O> query, final QueryOptions queryOptions) {
+        return new ResultSet<O>() {
+            @Override
+            public Iterator<O> iterator() {
+                scala.collection.Iterator<SimpleFeature> iter = getSimpleFeatureIteratorInternal((Intersects) query, queryOptions);
+
+                // JNH: Fix this:!?
+                return (Iterator<O>) JavaConversions.asJavaIterator(iter);
+            }
+
+            @Override
+            public boolean contains(O object) {
+                final Intersects intersects = (Intersects) query;
+                String attributeName = intersects.getAttributeName();
+                Geometry geom = (Geometry) object.getAttribute(attributeName);
+                return intersects.matchesValue(geom, queryOptions);
+            }
+
+            @Override
+            public boolean matches(O object) {
+                return query.matches(object, queryOptions);
+            }
+
+            @Override
+            public Query<O> getQuery() {
+                return query;
+            }
+
+            @Override
+            public QueryOptions getQueryOptions() {
+                return queryOptions;
+            }
+
+            @Override
+            public int getRetrievalCost() {
+                return INDEX_RETRIEVAL_COST;
+            }
+
+            // JNH: I'm returning the size here as the MergeCost.
+            // The geoindex size isn't optimal, so I don't know if this is good or not...
+            @Override
```

Comment:
We might want to evaluate this bit.  Not sure if there's a clean way to learn what these values should be.